### ISSUE TITLE
AssertionError executing RolloverFileOutputStreamTest 

### ISSUE THREAD

**Issue by lluismf** [eclipse/jetty.project#2120]
Building Jetty I get an assertion error in RolloverFileOutputStreamTest

[ERROR] testRolloverBackup(org.eclipse.jetty.util.RolloverFileOutputStreamTest)  Time elapsed: 10.02 s  <<< FAILURE!
java.lang.AssertionError: 

Expected: is <2>
     but: was <1>
	at org.eclipse.jetty.util.RolloverFileOutputStreamTest.testRolloverBackup(RolloverFileOutputStreamTest.java:352)

Maven Apache Maven 3.5.2
Maven home: c:\Program Files\apache-maven-3.5.2\bin\..
Java version: 1.8.0_112, vendor: Oracle Corporation
Java home: c:\Program Files\Java\jdk1.8.0_112\jre
Default locale: es_ES, platform encoding: Cp1251
OS name: "windows 7", version: "6.1", arch: "amd64", family: "windows"
![2018-01-15 00_56_23-workspace - java - jetty-util_src_test_java_org_eclipse_jetty_util_rolloverfileo](https://user-images.githubusercontent.com/2213724/34922357-166cbe06-f98f-11e7-8fd6-c2fef03e29de.png)

**Comment by lluismf** [comment]
The folder contains a single file test-rofos.log with contents BEFOREAFTER. 

![2018-01-15 01_24_37-workspace - debug - jetty-util_src_test_java_org_eclipse_jetty_util_rolloverfile](https://user-images.githubusercontent.com/2213724/34922647-04b3474e-f993-11e7-8feb-773a2eaaa747.png)

**Comment by gregw** [comment]
which branch?

**Comment by lluismf** [comment]
9.4.x

This line is the culprit (file can't be renamed, returns false). It could be a Windows only problem. 

https://github.com/eclipse/jetty.project/blob/21365234f8cf036829f7a03e5c7b89a9572444f9/jetty-util/src/main/java/org/eclipse/jetty/util/RolloverFileOutputStream.java#L283

**Comment by gregw** [comment]
Ah indeed, I think windows doesn't let a file be renamed while it is open.
There must be a way to switch files under windows... researching....

**Comment by gregw** [comment]
The class is fully synchronized, so no need to rename while open.  @lluismf  Can you try the fix in https://github.com/eclipse/jetty.project/tree/jetty-9.4.x-2120-rollover-on-windows

**Comment by joakime** [comment]
@gregw still has issue.

```
[INFO] Running org.eclipse.jetty.util.RolloverFileOutputStreamTest
[ERROR] Tests run: 8, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 10.112 s <<< FAILURE! - in org.eclipse.jetty.util.RolloverFileOutputStreamTest
[ERROR] testRollover(org.eclipse.jetty.util.RolloverFileOutputStreamTest)  Time elapsed: 0.011 s  <<< FAILURE!
java.lang.AssertionError: Unable to delete file: C:\code\jetty\jetty.project-master\jetty-util\target\tests\test-org.eclipse.jetty.util.RolloverFileOutputStreamTest_testRollover\test-rofos.log
        at org.eclipse.jetty.util.RolloverFileOutputStreamTest.testRollover(RolloverFileOutputStreamTest.java:291)
Caused by: java.nio.file.FileSystemException:
C:\code\jetty\jetty.project-master\jetty-util\target\tests\test-org.eclipse.jetty.util.RolloverFileOutputStreamTest_testRollover\test-rofos.log: The process cannot access the file because it is being used by another process.

        at org.eclipse.jetty.util.RolloverFileOutputStreamTest.testRollover(RolloverFileOutputStreamTest.java:291)

[INFO] Running org.eclipse.jetty.util.ScannerTest
```

**Comment by lluismf** [comment]
I don't know how the rollover with backup needs to work (or the rollover in general), is there any documentation to check? 
I guess the "old" outputstream must be closed before renaming/deleting the previous log.

**Comment by joakime** [comment]
@gregw this was a test locally, on my windows machine, of just the RolloverFileOutputStreamTest in jetty-util on your branch jetty-9.4.x-2120-rollover-on-windows

Windows is unique in its file locking.
You cannot close a file then delete it or rename it immediately as the next step in 100% of cases.
Even if you do a [`java.io.FileDescriptor.sync()`](https://docs.oracle.com/javase/8/docs/api/java/io/FileDescriptor.html#sync--) there will be a a period of time before you can delete or rename that file.

**Comment by joakime** [comment]
Supposedly [`java.nio.file.Files.move()`](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#move-java.nio.file.Path-java.nio.file.Path-java.nio.file.CopyOption...-) is safer to use, but I haven't tested that myself.

**Comment by joakime** [comment]
@gregw some interesting reading of `File.renameTo()` issues

* https://bugs.java.com/view_bug.do?bug_id=6213298 (workaround was to force a gc and then sleep for a bit of time)
* https://github.com/apache/commons-io/blob/master/src/main/java/org/apache/commons/io/FileUtils.java#L3005-L3013 (fix is to copy/delete if renameTo() fails.  but delete might still fail.)
* https://access.redhat.com/solutions/1414363 (solution is to copy file / delete, schedule delete if initial delete fails, don't reuse filename)
* https://bugs.java.com/bugdatabase/view_bug.do?bug_id=4073756 (fix is to copy / delete)

**Comment by lluismf** [comment]
Now two tests are failing:

- testRolloverBackup

java.lang.AssertionError: 
Expected: is "BEFORE"
     but: was "AFTER"

- testRollover

java.lang.AssertionError: Unable to delete file: C:\Users\lluis\git\jetty.project\jetty-util\target\tests\test-org.eclipse.jetty.util.RolloverFileOutputStreamTest_testRollover\test-rofos.log
...	at org.eclipse.jetty.toolchain.test.FS.deleteFile(FS.java:147)
	... 29 more

I think the file is locked by testRolloverBackup, when testRollover tries to empty the folder.

**Comment by lluismf** [comment]
Files.move also fails:

018-01-15 21:54:43.456:INFO::main: Logging initialized @349ms to org.eclipse.jetty.util.log.StdErrLog
java.nio.file.FileSystemException: C:\Users\lluis\git\jetty.project\jetty-util\target\tests\test-org.eclipse.jetty.util.RolloverFileOutputStreamTest_testRollover\test-rofos.log -> C:\Users\lluis\git\jetty.project\jetty-util\target\tests\test-org.eclipse.jetty.util.RolloverFileOutputStreamTest_testRollover\test-rofos.log.075448464: El proceso no tiene acceso al archivo porque está siendo utilizado por otro proceso.

**Comment by gregw** [comment]
BLAH!  How does anybody write meaningful applications on windows if renaming a file reliably is beyond the capabilities of that OS?!!?!?

I think we will just disable the backup mechanism for windows and ignore those tests.

**Comment by gregw** [comment]
So I've pushed an update. I use Files.move instead of rename, but if that fails the response is simply to give up on backup and instead append the to existing file (so no logs are  lost).

I also assume not windows for the backup test.

**Comment by joakime** [comment]
The issue with not being able to rename / move immediately after closing the file is not unique to Windows.
It can happen on network filesystems (NFS, SAN, CIFS, Samba, etc), slow filesystems (HPFS, older HDDs, etc), and also filesystems with metadata (OSX APFS, NTFS, etc).

That's why other open source projects that do file move/rename (apache commons-io, logback, log4j, etc) go for a tiered approach, do the simplest first (move), if that fails, the next most complicated (copy/delete), if that fails, schedule the delete (and don't reuse filename), if that fails, present the error/warning.

**Comment by joakime** [comment]
Closing, fixed (hopefully) in 8515f809caff57278b8eee48c6b95462805e0bba
